planner/core/rule_join_elimination.go

```diff
@@ -0,0 +1,184 @@
+	schemas []*expression.Schema
+}
+
+// tryToEliminateOuterJoin will eliminate outer join plan base on the following rules
+// 1. outer join elimination: For example left outer join, if the parent only use the
+//    columns from left table and the join key of right table(the inner table) is a unique
+//    key of the right table. the left outer join can be eliminated.
+// 2. outer join elimination with duplicate agnostic aggregate functions: For example left outer join.
+//    If the parent only use the columns from left table with 'distinct' label. The left outer join can
+//    be eliminated.
+func (o *outerJoinEliminator) tryToEliminateOuterJoin(p *LogicalJoin) LogicalPlan {
+	switch p.JoinType {
+	case LeftOuterJoin:
+		return o.doEliminate(p, 1)
+	case RightOuterJoin:
+		return o.doEliminate(p, 0)
+	default:
+		return nil
+	}
+}
+
+func (o *outerJoinEliminator) doEliminate(p *LogicalJoin, innerChildIdx int) LogicalPlan {
+	// outer join elimination with duplicate agnostic aggregate functions
+	if len(o.cols) > 0 {
+		cols := o.cols[len(o.cols)-1]
+		allColsInSchema := true
+		for _, col := range cols {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := p.children[1^innerChildIdx].Schema().FindColumn(columnName); c == nil {
+				allColsInSchema = false
+				break
+			}
+		}
+		if allColsInSchema == true {
+			return p.children[1^innerChildIdx]
+		}
+	}
+
+	// outer join elimination without duplicate agnostic aggregate functions
+	// first, check whether the parent's schema columns are all in left or right
+	if len(o.schemas) == 0 {
+		return nil
+	}
+	for _, col := range o.schemas[len(o.schemas)-1].Columns {
+		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+		if c, _ := p.children[1^innerChildIdx].Schema().FindColumn(columnName); c == nil {
+			return nil
+		}
+	}
+	// second, check whether the other side join keys are unique keys
+	p.children[innerChildIdx].buildKeyInfo()
+	var joinKeys []*expression.Column
+	for _, eqCond := range p.EqualConditions {
+		joinKeys = append(joinKeys, eqCond.GetArgs()[innerChildIdx].(*expression.Column))
+	}
+	tmpSchema := expression.NewSchema(joinKeys...)
+	for _, keyInfo := range p.children[innerChildIdx].Schema().Keys {
+		joinKeysContainKeyInfo := true
+		for _, col := range keyInfo {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := tmpSchema.FindColumn(columnName); c == nil {
+				joinKeysContainKeyInfo = false
+				break
+			}
+		}
+		if joinKeysContainKeyInfo {
+			return p.children[1^innerChildIdx]
+		}
+	}
+	// Third, if p.children[innerChildIdx] is datasource, we must check specially index.
+	// Because buildKeyInfo() don't save the index without notnull flag.
+	// But in outer join, null==null don't return true. The null index do not affect the res.
+	if ds, ok := p.children[innerChildIdx].(*DataSource); ok {
```

Comment:
This part is to check whether join keys contain a unique key set. But if an index can be null. It's not a unique key set because of two rows with NULL index are ok. However, in a left outer join, NULL compare everything is false. So it doesn't matter to consider NULL index.